Add message to CodeSmell

Mauin opened the pull request:
This is just a suggestion and currently won't compile as we'd have to update every usage of `CodeSmell`.

After the discussions in #475 and #273 it seemed to me like splitting an Issues description from a CodeSmell's detailed message makes sense.

With this the Issue description can act as a rationale for the Issue, what it means, the reasoning behind the rule and how it can be fixed. (e.g. "Long function names should be avoided because.......<insert long and detailed description and reasoning here>")

The CodeSmell message can then point out the specific problem directly. (e.g. "The function ${function.name} has a ${function.name.lenght} character long name which is longer than the threshold of $maxFunctionLength. Give this function a shorter name")

Let me know what you think. If you agree with this I can go through all places we report CodeSmells and give them an empty message to start with `""` and we should update them over time.

arturbosch commented:
So the `issue.description` is the generic message and the `CodeSmell.message` is the concrete? I like it. It will remove `issue.copy` :+1:

Mauin commented:
@arturbosch Exactly. Then I will go ahead and add the `CodeSmell.message` everywhere as well

Mauin commented:
I now added a `message` to every constructor of `CodeSmell` (thanks IntelliJ for doing almost all the work!!). 

In most cases this detailed message is just `""` and we should add a more detailed message later. I didn't want to go through every `CodeSmell` we report and think of a message that seemed logical to me.. 🙃 

Also I moved all cases where we did `issue.copy(description = "...")` and took that detailed message as the new `CodeSmell#message`.

arturbosch commented:
Argh just came by to say "hold on until my PR is through!" :D. Thats a huge PR and will conflict with mine. I refactor the `MultiRule`. But no worries we will get both things merged soon.

Mauin commented:
@arturbosch Whoooooops 🙃 

But the changes in here are fairly "simple" to rebase, so don't worry too much. Mostly just a lot of `message = ""` everywhere.

arturbosch commented:
I want to wait with this PR after RC5 release (hopefully soon).

Mauin commented:
@arturbosch Now that RC5 is out, should I get started on rebasing this and getting all these conflicts resolved?

arturbosch commented:
Yes, please. 

I don't think we should print the description of each finding. This will fill up the console and is printed after the findings path. Having the description in the output format is enough imo.

schalkms commented:
I agree with @arturbosch 
Detekt should not fill up the console.

Mauin commented:
Finally got around to rebasing this. This change currently won't print the `message` of each finding in the CLI output. It will use either the `message` in the XML output if present, otherwise it falls back to the `issue#description`.

arturbosch commented:
Great! Will check it out tomorrow.

arturbosch commented on detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/CodeSmell.kt:
```diff
@@ -27,23 +33,27 @@ open class CodeSmell(final override val issue: Issue,
+		issue, entity, metric, message, references = listOf(reference)) {
 
-	override fun compact() = "$id - $metric - ref=${reference.name} - ${entity.compact()}"
+	override fun compact() = "$id - $metric - ref=${reference.name} - ${entity.compact()} - message=$message"
```
This still gets printed in the console

arturbosch commented on detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/CodeSmell.kt:
```diff
@@ -27,23 +33,27 @@ open class CodeSmell(final override val issue: Issue,
 		get() = metric.threshold
 
-	override fun compact() = "$id - $metric - ${entity.compact()}"
+	override fun compact() = "$id - $metric - ${entity.compact()} = message=$message"
```
Here too

arturbosch commented on detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/out/XmlOutputReport.kt:
```diff
@@ -30,11 +30,16 @@ class XmlOutputReport : OutputReport() {
 		smells.groupBy { it.location.file }.forEach { fileName, findings ->
 			lines += "<file name=\"${fileName.toXmlString()}\">"
 			findings.forEach {
+				val message = when {
```
I would move that logic into the CodeSmell class and use a function like `messageOrDescription`. It feels a bit like feature envy.

Mauin replied:
I'm hoping that at some point all `CodeSmell`s will just have a valid message so we can remove this again. But for now it makes sense to at least print whatever is there.

arturbosch commented:
Thank you!
